src/lua_test.c

```diff
@@ -0,0 +1,151 @@
+/**
+ * Copyright (C) 2024 Kentaro Hayashi
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * Authors:
+ *   Kentaro Hayashi <kenhys at xdump.org>
+ **/
+
+#define TEST_PLUGIN_LUA 1
+
+#include "lua.c"
+#include "testing.h"
+#include "utils_lua.c"
+#include "utils_lua.h"
+
+#define TEST_LUA_PLUGIN_CONF "/tmp/collectd_lua_plugin.conf"
+
+static lua_State *L = NULL;
+
+void test_setup_config(const char *config) {
+  FILE *fp = fopen(TEST_LUA_PLUGIN_CONF, "w+");
+  fprintf(fp, "%s", config);
+  fclose(fp);
+}
+
+void test_setup(void) {
+  L = luaL_newstate();
+  luaL_openlibs(L);
+}
+
+void test_teardown(void) {
+  unlink(TEST_LUA_PLUGIN_CONF);
+  lua_close(L);
+}
+
+const char *module_key_value_config = "<Plugin lua>\n"
+                                      "  <Module>\n"
+                                      "    Key \"Value\"\n"
+                                      "  </Module>\n"
+                                      "</Plugin>\n";
+
+DEF_TEST(plugin_config) {
+  /*
+    <root>
+      <Plugin lua>
+        ...
+      </Plugin>
+    </root>
+  */
+  test_setup();
+  test_setup_config(module_key_value_config);
+  oconfig_item_t *root = oconfig_parse_file(TEST_LUA_PLUGIN_CONF);
+  oconfig_item_t *plugin, *module, *ci;
+  char buf[PATH_MAX];
+  CHECK_NOT_NULL(root);
+  EXPECT_EQ_INT(0, root->values_num);
```

Comment:
It is added to ensure oconfig_parse_file behavior, but also redundant in this context. 
I've removed it.

Reply:
> As it is, this test verifies that the config parser works as advertised. This is very valuable – the config parsing is woefully undertested. Untested, really.

@octo, for Sysman plugin I have a directory with invalid collectd config files, which are parsed with collectd under Valgrind, with the expectation that collectd should return error for each config, but Valgrind should not report any errors.

Similar approach works as well for testing basic collectd config handling and any other plugin:
```
# test configs are supposed to be in same dir as this script
confdir=${0%/*}

failexit=1
memexit=33

fail=0
unknown=0
memcheck=0
output=$(mktemp)
testconf=$(mktemp)
echo "Validating invalid config handling..."
for conf in "$confdir"/*.cnf; do
        echo
        echo "$conf:"
        cat "$baseconf" "$conf" > "$testconf"
        valgrind --error-exitcode=$memexit "$collectd" -T -C "$testconf" > "$output" 2>&1
        ret=$?
        if [ $ret -eq $failexit ]; then
                echo "  PASS: exit=$ret"
                fail=$((fail+1))
        elif [ $ret -eq $memexit ]; then
                echo "  ERROR: unexpected valgrind error"
                memcheck=$((memcheck+1))
        else
                echo "  ERROR: unexpected collectd exit=$ret"
                unknown=$((unknown+1))
        fi
        echo $line
        cat "$output"
        echo $line
done
rm "$baseconf" "$output"

echo
echo "Config parsing results:"
echo "- $fail expected fails"
echo "- $unknown unexpected exits"
echo "- $memcheck unexpected valgrind failures"
if [ $((unknown+memcheck)) -gt 0 ] ; then
        echo "=> ERROR"
        exit 1
fi
echo "=> PASS"
```

PS. Instead of using Valgrind, collectd could be built with ASAN as that's much faster than Valgrind, and checks also stack accesses.  In general, it might be good to run also other tests with ASAN built code: https://github.com/google/sanitizers/wiki/AddressSanitizer